clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -182,31 +182,32 @@ public void done(long baseOffset, long logAppendTime, RuntimeException exception
         Iterator<MutableRecordBatch> recordBatchIter = memoryRecords.batches().iterator();
         if (!recordBatchIter.hasNext())
             throw new IllegalStateException("Cannot split an empty producer batch.");
-        RecordBatch recordBatch = recordBatchIter.next();
-        if (recordBatchIter.hasNext())
-            throw new IllegalStateException("A producer batch should only have one record batch.");
-
         Iterator<Thunk> thunkIter = thunks.iterator();
         // We always allocate batch size because we are already splitting a big batch.
         // And we also Retain the create time of the original batch.
-        ProducerBatch batch = null;
-        for (Record record : recordBatch) {
-            assert thunkIter.hasNext();
-            Thunk thunk = thunkIter.next();
-            if (batch == null) {
-                batch = createBatchOffAccumulatorForRecord(record, splitBatchSize);
+        while (recordBatchIter.hasNext()) {
```

Comment:
If the message format is v0 or v1, then we could have multiple batches (each record is a batch of size 1). Obviously the batch splitting is intended for batches with multiple records, but it felt a little awkward and unnecessary to restrict this function to only magic >= 2 or compression != NONE.